src/bp/core/services/dialog/decision-engine.ts

```diff
@@ -25,75 +31,102 @@ export class DecisionEngine {
   ) {}
 
   private readonly MIN_CONFIDENCE = process.env.BP_DECISION_MIN_CONFIENCE || 0.3
+  private readonly MIN_NO_REPEAT = ms(process.env.BP_DECISION_MIN_NO_REPEAT || '20s')
 
   public async processEvent(sessionId: string, event: IO.IncomingEvent) {
     const isInMiddleOfFlow = _.get(event, 'state.context.currentFlow', false)
+    if (!event.suggestions) {
+      Object.assign(event, { suggestions: [] })
+    }
 
-    if (event.suggestions && !isInMiddleOfFlow) {
-      const bestReply = this._findBestReply(event)
-      if (bestReply) {
-        await this._sendSuggestion(bestReply, sessionId, event)
-      }
+    this._amendSuggestionsWithDecision(event.suggestions!, _.get(event, 'state.session.lastMessages', []))
+
+    if (isInMiddleOfFlow) {
+      event.suggestions!.forEach(suggestion => {
+        if (suggestion.decision.status === 'elected') {
+          suggestion.decision.status = 'dropped'
+          suggestion.decision.reason = 'would have been elected, but already in the middle of a flow'
+        }
+      })
+    }
+
+    if (this.onBeforeSuggestionsElection) {
+      await this.onBeforeSuggestionsElection(sessionId, event, event.suggestions!)
+    }
+
+    const elected = event.suggestions!.find(x => x.decision.status === 'elected')
+
+    if (elected) {
+      Object.assign(event, { decision: elected })
+      await this._sendSuggestion(elected, sessionId, event)
     }
 
     if (!event.hasFlag(WellKnownFlags.SKIP_DIALOG_ENGINE)) {
       try {
+        Object.assign(event, {
```

Comment:
what's this case ? isn't it suppose to ne the case where no suggestion match ? it seems like there is no check and will always be marked as elected except when skip dialog engine is true. I'm probably missing somth

Reply:
This will be triggered whenever no module has anything to suggest, which is the case when the QnA doesn’t suggest a reply.